include/ParticlesData.hpp

```diff
@@ -63,6 +124,141 @@ struct ParticlesData
     const static T K;
 };
 
+template <typename T>
+struct ParticlesData
+{
+    inline void resize(const size_t size)
+    {
+        for (unsigned int i = 0; i < data.size(); ++i)
+            data[i]->resize(size);
+    }
+
+    size_t iteration;                            // Current iteration
+    size_t n, side, count;                       // Number of particles
+    std::vector<T> x, y, z, x_m1, y_m1, z_m1;    // Positions
+    std::vector<T> vx, vy, vz;                   // Velocities
+    std::vector<T> ro;                           // Density
+    std::vector<T> u;                            // Internal Energy
+    std::vector<T> p;                            // Pressure
+    std::vector<T> h;                            // Smoothing Length
+    std::vector<T> m;                            // Mass
+    std::vector<T> c;                            // Speed of sound
+    std::vector<T> grad_P_x, grad_P_y, grad_P_z; // gradient of the pressure
+    std::vector<T> du, du_m1;                    // variation of the energy
+    std::vector<T> dt, dt_m1;
+    std::vector<T> c11, c12, c13, c22, c23, c33; // IAD components
+    std::vector<T> maxvsignal;
+
+    // For Sedov
+    std::vector<T> mui, temp, cv;
+
+    T ttot, etot, ecin, eint;
+    T minDt;
+
+    BBox<T> bbox;
+
+    std::vector<std::vector<T> *> data{&x,   &y,   &z,   &x_m1, &y_m1, &z_m1,     &vx,         &vy,       &vz,   &ro,    &u,
+                                       &p,   &h,   &m,    &c,    &grad_P_x, &grad_P_y,   &grad_P_z, &du,   &du_m1, &dt,   &dt_m1,
+                                       &c11, &c12, &c13, &c22,  &c23,  &c33,      &maxvsignal, &mui,      &temp, &cv};
+
+    const std::array<double, lt::size> wh = lt::createWharmonicLookupTable<double, lt::size>();
+    const std::array<double, lt::size> whd = lt::createWharmonicDerivativeLookupTable<double, lt::size>();
+
+#ifdef USE_MPI
+    MPI_Comm comm;
+    int pnamelen = 0;
+    char pname[MPI_MAX_PROCESSOR_NAME];
+#endif
+
+
+#if defined(USE_CUDA)
+    // number of CUDA streams to use
+    static const int NST = 2;
+
+    cudaStream_t streams[NST];
+    
+    int *d_clist[NST], *d_neighbors[NST], *d_neighborsCount[NST]; // work arrays per stream
```

Comment:
the cuda streams and the d_clist, d_neighbors and d_neighborsCount could go into a separate
struct `Workspace`. Then you can do `Workspace streamScratch[NST]`. Each workspace takes
care of allocation and deallocation.